src/plugins/intel_cpu/src/nodes/mkldnn_topk_node.cpp

```diff
@@ -42,378 +1823,499 @@ bool MKLDNNTopKNode::isSupportedOperation(const std::shared_ptr<const ngraph::No
-        int tmp_index;
-        int s_index = i0 * dim * after_num + first_index + i1;
 
-        auto swap_func = [&](int index1, int index2) {
-            tmp_value = max_values[index1];
-            max_values[index1] = max_values[index2];
-            max_values[index2] = tmp_value;
+        if (compile_kernel) {
+            auto jcp = jit_topk_config_params();
+            jcp.precision = selectedPD->getConfig().inConfs[TOPK_DATA].desc->getPrecision();
+            jcp.data_size = data_size;
+            jcp.blk_size = blk_size;
+            jcp.layout = layout;
+            jcp.top_k = top_k;
+            jcp.axis_dim = axis_dim;
+            jcp.mode_max = mode_max;
+            jcp.sort_index = sort_index;
+            jcp.topk_innermost = topk_innermost;
+            jcp.algorithm = algorithm;
+            jcp.bubble_inplace = bubble_inplace;
+            jcp.sort_stride = static_cast<int>(I);
+            jcp.work_amount = static_cast<int>(I);
 
-            tmp_index = max_indexes[index1];
-            max_indexes[index1] = max_indexes[index2];
-            max_indexes[index2] = tmp_index;
-        };
+            if (algorithm == TopKAlgorithm::topk_bitonic_sort) {
+                size_t src_count = srcMemPtr->GetDescWithType<BlockedMemoryDesc>()->getPaddedElementsCount();
+                vec_process_ptr.resize(src_count * data_size);
+                vec_process_idx_ptr.resize(src_count * sizeof(int32_t));
 
-        for (int i2 = 0; i2 < src_k; i2++) {
-            max_values[i2] = src_data[s_index];
-            max_indexes[i2] = i2;
-            s_index += after_num;
-        }
-        for (int i2 = 0; i2 < src_k - 1; i2++) {
-            for (int i3 = src_k - 1; i3 > i2; i3--) {
-                if (Compare2<float>()(max_values[i3], max_values[i3 - 1])) {
-                    swap_func(i3, i3 - 1);
+                calc_bitonic_idx(axis_dim, jcp.bitonic_idx_cnt, true);
+                if (sort_index) {
+                    calc_bitonic_idx(top_k, jcp.bitonic_k_idx_cnt, false);
                 }
             }
-        }
-        for (int i2 = src_k; i2 < dim; i2++) {
-            max_values[src_k] = src_data[s_index];
-            max_indexes[src_k] = i2;
-            for (int i3 = src_k; i3 > 0; i3--) {
-                if (Compare2<float>()(max_values[i3], max_values[i3 - 1]))
-                    swap_func(i3, i3 - 1);
-                else
-                    break;
+
+            if (mayiuse(cpu::x64::avx512_common)) {
+                topk_kernel.reset(new jit_uni_topk_kernel_f32<cpu::x64::avx512_common>(jcp));
+            } else if (mayiuse(cpu::x64::avx2)) {
+                topk_kernel.reset(new jit_uni_topk_kernel_f32<cpu::x64::avx2>(jcp));
+            } else if (mayiuse(cpu::x64::sse41)) {
+                topk_kernel.reset(new jit_uni_topk_kernel_f32<cpu::x64::sse41>(jcp));
             }
-            s_index += after_num;
+
+            if (topk_kernel)
+                topk_kernel->create_ker();
```

Comment:
In case of dynamic shapes prepareParams is called on the first inference. In our words even if jit_kernel is shape agnostic we create it on inference stage and as follows affect first inference latency. Is it possible to move all preparation steps which doesn't depend on dynamic params in CreatePrimitive routine?

Reply:
Great suggestion. Fixed. Thanks Dmitry!

Reply:
> @xuchen-intel could you please also check that TopK tests pass for sse41 ISA using https://github.com/openvinotoolkit/openvino/wiki/CPUPluginIntelSoftwareDevelopmentEmulator?

Done. Check passed. Check results have been sent. Thanks Dmitry!

Reply:
Besides applying the review comments, I've also added some code comments, function renaming and reuse, compared to Wednesday's update, to improve the code readability.

Reply:
> > @xuchen-intel could you please also check that TopK tests pass for sse41 ISA using https://github.com/openvinotoolkit/openvino/wiki/CPUPluginIntelSoftwareDevelopmentEmulator?
> 
> Done. Check passed. Check results have been sent. Thanks Dmitry!

Test passed for both -ast and -nhm. Thanks Dmitry!